way-cooler/src/shells/xdg.rs

```diff
@@ -100,6 +100,59 @@ impl XdgShellHandler for Xdg {
         }).unwrap();
     }
 
+    fn map_request(&mut self,
+                   compositor: CompositorHandle,
+                   _: SurfaceHandle,
+                   shell_surface_handle: XdgShellSurfaceHandle) {
+        let is_toplevel = with_handles!([(shell_surface: {&shell_surface_handle})] => {
+            match shell_surface.state().unwrap() {
+                TopLevel(_) => true,
+                _ => false
+            }
+        }).unwrap();
```

Comment:
`dehandle!` returns a value, so you can instead do:

```rust
        let is_toplevel = dehandle!(
            @shell_surface = {&shell_surface_handle};
            match shell_surface.state().unwrap() {
                TopLevel(_) => true,
                _ => false
            }
        );
```

and there's no need to do the ugly `unwrap()` at the end.

Can you also put a comment about the shell surface handle being used in `seat.focus_view` which is why it can't be combined with the next `dehandle!` block? I rant into that pitfall just now while reviewing your code.